transcoders.go

```diff
@@ -238,13 +238,66 @@ func garlic64BtS(b []byte) (string, error) {
 	return addr, nil
 }
 
-func garlicValidate(b []byte) error {
+func garlic64Validate(b []byte) error {
 	if len(b) < 386 {
 		return fmt.Errorf("failed to validate garlic addr: %s not an i2p base64 address. len: %d\n", b, len(b))
 	}
 	return nil
 }
 
+var TranscoderGarlic32 = NewTranscoderFromFunctions(garlic32StB, garlic32BtS, garlic32Validate)
+
+var garlicBase32Encoding = base32.NewEncoding("abcdefghijklmnopqrstuvwxyz234567")
+
+func garlic32StB(s string) ([]byte, error) {
+	s = strings.Replace(s, ".b32.i2p", "", -1)
+	// garlic address without the ".b32.i2p" substring
+	if len(s) < 55 {
+		if len(s) != 52 {
+			return nil, fmt.Errorf("failed to parse garlic addr: %s not a i2p base32 address. len: %d", s, len(s))
+		}
+	}
+
+	padout := func(s2 string) string {
+		str := ""
+		for x := 0; x < 56; x++ {
+			if x < len(s) {
+				str += string(s[x])
+			} else {
+				str += "="
+			}
+		}
+		return str
+	}
+
+	garlicHostBytes, err := garlicBase32Encoding.DecodeString(padout(s))
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode base32 garlic addr: %s err %s len %v", s, err, len(s))
+	}
+	return garlicHostBytes, nil
+}
+
+func garlic32BtS(b []byte) (string, error) {
+	if len(b) > 35 {
+		if len(b) < 32 {
+			return "", fmt.Errorf("failed to validate garlic addr: %s not an i2p base32 address. len: %d\n", b, len(b))
+		}
+	}
+	unpad := func(s2 string) string {
+		return strings.Replace(s2, "=", "", -1)
```

Comment:
`strings.TrimRight`?

Reply:
Yeah not sure where my head was at here. Fixed.